x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.support;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.IdsQueryBuilder;
+import org.elasticsearch.index.query.MatchAllQueryBuilder;
+import org.elasticsearch.index.query.PrefixQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.RangeQueryBuilder;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.index.query.TermQueryBuilder;
+import org.elasticsearch.index.query.TermsQueryBuilder;
+import org.elasticsearch.index.query.WildcardQueryBuilder;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.security.authc.ApiKeyService;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+
+public class ApiKeyBoolQueryBuilder extends BoolQueryBuilder {
+
+    // Field names allowed at the index level
+    private static final Set<String> ALLOWED_EXACT_INDEX_FIELD_NAMES =
+        Set.of("doc_type", "name", "api_key_invalidated", "creation_time", "expiration_time");
+
+    private ApiKeyBoolQueryBuilder() {}
+
+    public static ApiKeyBoolQueryBuilder build(QueryBuilder queryBuilder, Authentication authentication) {
+        final ApiKeyBoolQueryBuilder finalQuery = new ApiKeyBoolQueryBuilder();
+        if (queryBuilder != null) {
+            QueryBuilder processedQuery = doProcess(queryBuilder);
+            if (false == processedQuery instanceof BoolQueryBuilder) {
+                finalQuery.must(processedQuery);
+            } else {
+                final BoolQueryBuilder boolQueryBuilder = (BoolQueryBuilder) processedQuery;
+                finalQuery.minimumShouldMatch(boolQueryBuilder.minimumShouldMatch());
+                finalQuery.adjustPureNegative(boolQueryBuilder.adjustPureNegative());
+                boolQueryBuilder.must().forEach(finalQuery::must);
+                boolQueryBuilder.should().forEach(finalQuery::should);
+                boolQueryBuilder.mustNot().forEach(finalQuery::mustNot);
+                boolQueryBuilder.filter().forEach(finalQuery::filter);
+            }
```

Comment:
Is this special case necessary? Isn't it simpler to always wrap it in a new bool so we know everything works the same way?

Reply:
The intention is to always have the final query to be an instance of `ApiKeyBoolQueryBuilder` because it has special handling to disable certain fields on the SearchExecutionContext level. So yes this is necessary because it converts a normal boolQueryBuilder to a `ApiKeyBoolQueryBuilder`.

Reply:
But can't we just have the `if` case used all the time? Why doesn't that do what we want?
```
           QueryBuilder processedQuery = doProcess(queryBuilder);
           finalQuery.must(processedQuery);
```

Reply:
Ah ok. Sorry I see what you mean now. You are right! The must clause can simply accept any query builder including a bool query (now it's kinda obvious after you pointed it out ...) It's a better approach and I assume the end results are the same after lucene level query rewrite. Will update. Thanks!